src/core/speed.h

```diff
@@ -0,0 +1,73 @@
+#ifndef CORE_SPEED_H
+#define CORE_SPEED_H
+
+#include "core/time.h"
+
+/**
+ * @file
+ * Scroll speed functions.
+ * Calculates deltas for constant speed/acceleration/deceleration based on
+ * a "number of pixels per unit of time" value.
+ */
+
+/**
+ * Constant for immediate speed change with no acceleration/deceleration
+ */
+#define SPEED_CHANGE_IMMEDIATE 0
+
+/**
+* Speed direction constants
+*/
+typedef enum {
+    SPEED_DIRECTION_NEGATIVE = -1,
+    SPEED_DIRECTION_STOPPED  = 0,
+    SPEED_DIRECTION_POSITIVE = 1
+} speed_direction;
+
+typedef struct {
+    time_millis start_time;
+    time_millis total_time;
+    time_millis last_speed_check;
+    int speed_difference;
+    int desired_speed;
+    int current_speed;
+    int cumulative_delta;
+} speed;
```

Comment:
Call this `speed_type` instead. That way you can name the function arguments/variables just `speed` instead of `sp`, making the code clearer.